predicators/main.py

```diff
@@ -393,12 +397,82 @@ def _run_testing(env: BaseEnv, approach: BaseApproach) -> Metrics:
             "num_nodes_created", "num_nsrts", "num_preds", "plan_length",
             "num_failures_discovered"
     ]:
-        total = approach.metrics[f"total_{metric_name}"]
+        total = cogman.metrics[f"total_{metric_name}"]
         metrics[f"avg_{metric_name}"] = (
             total / num_found_policy if num_found_policy > 0 else float("inf"))
     return metrics
 
 
+def _run_episode(
+    cogman: CogMan,
+    env: BaseEnv,
+    train_or_test: str,
+    task_idx: int,
+    max_num_steps: int,
+    do_env_reset: bool = True,
+    exceptions_to_break_on: Optional[Set[TypingType[Exception]]] = None,
+    monitor: Optional[utils.Monitor] = None
+) -> Tuple[Tuple[List[Observation], List[Action]], bool, Metrics]:
+    """Execute cogman starting from the initial state of a train or test task
+    in the environment.
+
+    Note that the environment and cogman internal states are updated.
+
+    Terminates when any of these conditions hold:
+    (1) the termination_function returns True
+    (2) max_num_steps is reached
+    (3) cogman or env raise an exception of type in exceptions_to_break_on
+
+    Note that in the case where the exception is raised in step, we exclude the
+    last action from the returned trajectory to maintain the invariant that
+    the trajectory states are of length one greater than the actions.
+
+    This is defined here mostly to avoid circular import issues for cogman.
+    We may want to move it eventually.
+    """
+    if do_env_reset:
+        env.reset(train_or_test, task_idx)
+    obs = env.get_observation()
+    observations = [obs]
+    actions: List[Action] = []
+    metrics: Metrics = defaultdict(float)
+    metrics["policy_call_time"] = 0.0
+    exception_raised_in_step = False
+    if not env.goal_reached():
+        for _ in range(max_num_steps):
+            monitor_observed = False
+            exception_raised_in_step = False
+            try:
+                start_time = time.perf_counter()
+                act = cogman.step(obs)
+                metrics["policy_call_time"] += time.perf_counter() - start_time
+                # Note: it's important to call monitor.observe() before
+                # env.step(), because the monitor may use the environment's
+                # internal state.
```

Comment:
Isn't this 'cheating'? Are we planning to eventually move away from this? If so, would be good to note that here.

Reply:
kinda, but the monitor isn't part of the agent, it's just meant to collect statistics and stuff like that. so it doesn't really matter what it accesses

Reply:
hmmm - that's fair. When we think of actually using a robot tho, wouldn't the monitor have to use a perceiver or something like that?

Reply:
good point. okay after thinking about it more, I think the original comment is actually misleading, and I was wrong to say that the monitor is cheating. see the updated comment

Reply:
nice - i like it!

Reply:
wait - i'm kinda lost now; how come the `BaseExecutionMonitor` doesn't implement an `observe()` method?

Reply:
ohhh I understand the confusion. this is talking about `Monitor` https://github.com/Learning-and-Intelligent-Systems/predicators/tree/master/predicators#L913 not the new execution monitor stuff 😅

Reply:
Wait - given this, can we do some renaming to avoid this confusion going forward? We could call the current `Monitor` something like `LoggingMonitor`? Or we could rename `Execution Monitor` to `ExMon` in keeping with the `CogMan` convention?

Reply:
I changed `Monitor` to `LoggingMonitor`. I think `ExMon` would probably be more confusing, so I'll vote for keeping `ExecutionMonitor`